feat(linter): Implement `no-unexpected-multiline` rule

camchenry opened the pull request:
- part of https://github.com/oxc-project/oxc/issues/479

The bulk of this rule is closely based on the logic from the original ESLint rule. One major difference between this implementation and the original though is the lack of a tokenizer. ESLint uses a proper tokenizer to find identifers, parentheses, brackets, and other tokens. For this rule, I opted to just manually search for the characters we might expect to find. I'm not sure how this will hold up in the real world, I expect it could lead to some missing cases potentially, but it at least works for all of the given test cases.

camchenry commented:
I still need to polish this, add more docs, and add better diagnostics, then will put up for review

DonIsaac commented on crates/oxc_linter/src/rules/eslint/no_unexpected_multiline.rs:
```diff
@@ -0,0 +1,369 @@
+                        span.start + u32::try_from(open_paren).unwrap(),
+                        span.start + u32::try_from(open_paren + 1).unwrap(),
+                    );
+                    let diagnostic = OxcDiagnostic::warn(
```
Please move diagnostic building to a separate function (at the top of the file). It makes quickly seeing what kind of messages this lint rule warns about much easier

DonIsaac commented on crates/oxc_linter/src/rules/eslint/no_unexpected_multiline.rs:
```diff
@@ -0,0 +1,369 @@
+                };
+                if newline < open_bracket {
+                    let bracket_span = Span::new(
+                        span.start + u32::try_from(open_bracket).unwrap(),
```
you may find `Span::sized` helpful

camchenry replied:
good to know!

DonIsaac commented on crates/oxc_linter/src/rules/eslint/no_unexpected_multiline.rs:
```diff
@@ -0,0 +1,369 @@
+                    .iter()
+                    .position(|&b| !b.is_ascii_alphanumeric())
+                {
+                    &src[second_slash + 1..second_slash + 1 + offset]
```
issue: I've seen ranged index on `source_text` cause rules to panic in the past (e.g. weird unicode characters cause this to end on a character boundary). Is it possible to refactor this logic to use `chars()` instead?

camchenry replied:
good idea, I wish we had a good way to quickly tokenize some text here, but I refactored this to at least do a good enough job at finding identifier-like strings

camchenry committed:
feat(linter): Implement `no-unexpected-multiline` rule (#5911)

- part of https://github.com/oxc-project/oxc/issues/479

The bulk of this rule is closely based on the logic from the original ESLint rule. One major difference between this implementation and the original though is the lack of a tokenizer. ESLint uses a proper tokenizer to find identifers, parentheses, brackets, and other tokens. For this rule, I opted to just manually search for the characters we might expect to find. I'm not sure how this will hold up in the real world, I expect it could lead to some missing cases potentially, but it at least works for all of the given test cases.
